segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,450 @@
+                        long length = chunkInfo.getLength();
+
+                        ChunkMetadata lastChunk = (ChunkMetadata) txn.get(segmentMetadata.getLastChunk());
+                        Preconditions.checkState(null != lastChunk);
+                        lastChunk.setLength(Math.toIntExact(length));
+                        txn.update(lastChunk);
+                        segmentMetadata.setLength(segmentMetadata.getLength() + length);
+                    }
+                    Preconditions.checkState(segmentMetadata.isOwnershipChanged());
+                    segmentMetadata.checkInvariants();
+                    txn.update(segmentMetadata);
+                }
+
+                // Step 4: Apply the truncate offsets.
+                for (String systemSegment : systemSegments) {
+                    if (finalTruncateOffsets.containsKey(systemSegment)) {
+                        val truncateAt = finalTruncateOffsets.get(systemSegment);
+                        val firstChunkStartsAt = finalFirstChunkStartsAtOffsets.get(systemSegment);
+                        applyTruncate(txn, systemSegment, truncateAt, firstChunkStartsAt);
+                    }
+                }
+
+                // Step 5: Finally commit all data.
+                txn.commit(true, true);
+            }
+        }
+    }
+
+    /**
+     * Apply chunk addition.
+     */
+    private void applyChunkAddition(MetadataTransaction txn, HashMap<String, Long> chunkStartOffsets, String segmentName, String oldChunkName, String newChunkName, long offset) throws StorageMetadataException {
+        Preconditions.checkState(null != oldChunkName);
+        Preconditions.checkState(null != newChunkName && !newChunkName.isEmpty());
+
+        SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(segmentName);
+        segmentMetadata.checkInvariants();
+
+        // set length.
+        segmentMetadata.setLength(offset);
+
+        val newChunkMetadata = ChunkMetadata.builder()
+                .name(newChunkName)
+                .build();
+        txn.create(newChunkMetadata);
+        txn.markPinned(newChunkMetadata);
+
+        chunkStartOffsets.put(newChunkName, offset);
+        // Set first and last pointers.
+        if (!oldChunkName.isEmpty()) {
+            ChunkMetadata oldChunk = (ChunkMetadata) txn.get(oldChunkName);
+            Preconditions.checkState(null != oldChunk);
+
+            // In case the old segment store was still writing some zombie chunks when ownership changed
+            // then new offset may invalidate tail part of chunk list.
+            // Note that chunk with oldChunkName is still valid, it is the chunks after this that become invalid.
+            String toDelete = oldChunk.getNextChunk();
+            while (toDelete != null) {
+                ChunkMetadata chunkToDelete = (ChunkMetadata) txn.get(toDelete);
+                txn.delete(toDelete);
+                toDelete = chunkToDelete.getNextChunk();
```

Comment:
do we need to get before we delete? I'm not clear that the order is correct.

Reply:
Yes. We are iterating over the list of chunks.  this line `ChunkMetadata chunkToDelete = (ChunkMetadata) txn.get(toDelete);` retrieved the record "pointed" by toDelete and the next line `toDelete = chunkToDelete.getNextChunk();` moves the pointer.  That is why we first get a reference to copy of the record and then delete it from the transaction. Once deleted we won't be able to find that record.